### PR TITLE
add param validation for list api's

### PR DESCRIPTION
**Pull request by cpirich**
- migrated a subset of the parameter validation functions to work with the special "don't marshal" APIs that operate directly on interpreter data structures
- connected basic param validation for `insertItem`, `appendItem`, and `removeItem`. Note that we do not validate the values populated in the lists - those are allowed to be anything - including undefined.
- added an extra warning to `removeItem` to cover the case when the index is too large.

### CODE REVIEW

**Comment by davidsbailey** [comment]
Thanks Chris, I'll take a look. Is this related to https://www.pivotaltracker.com/story/show/99825840 "calling Array.prototype.appendItem causes infinite loop of errors" ?

File: apps/src/applab/dontMarshalApi.js [hunk 1]
```diff
@@ -42,10 +83,17 @@ exports.removeItem = function (array, index) {
   if (index < array.length) {
     delete array.properties[array.length - 1];
     array.length -= 1;
+  } else {
+    // index is out of bounds (too large):
+    outputWarning("removeItem() index parameter value (" + index +
+                    ") is larger than the number of items in the list (" +
```
Comment (davidsbailey) [inline apps/src/applab/dontMarshalApi.js:89]: Perhaps we should have the same behavior for negative indices which are too large? Currently we delete the first element in that case, but it seems like doing nothing and giving a warning might be more appropriate as it most likely indicates a programming error.

File: apps/src/applab/dontMarshalApi.js [hunk 2]
```diff
@@ -1,9 +1,44 @@
...
 //
 // dropletConfig for each of these APIs should be marked with dontMarshal:true
 
+
+function dmapiValidateType(funcName, varName, varValue, expectedType, opt) {
+  var properType;
+  if (typeof varValue !== 'undefined') {
+    if (expectedType === 'number') {
+      properType = (typeof varValue.data === 'number' ||
+                    (typeof varValue.data === 'string' && !isNaN(varValue.data)));
```
Comment (davidsbailey) [inline apps/src/applab/dontMarshalApi.js:25]: it looks like validation will fail when `expectedType` is `string` and `varValue` is `'5'` (a string which can be parsed into a number). Why?

Comment (davidsbailey) [inline apps/src/applab/dontMarshalApi.js:25]: OK, I see that I misread this -- you are accepting numerical strings when expectedType is 'number'. Makes sense.

**Comment by davidsbailey** [comment]
LGTM after considering the inline suggestion. I see after reviewing that this is unrelated to the pivotal item i asked about.

File: apps/src/applab/dontMarshalApi.js (hunk 1 above)
Comment (cpirich) [inline apps/src/applab/dontMarshalApi.js:89]: I thought about that, but decided not to change the function's behavior. That might require a doc change and potentially break apps that were already written (thought I'd expect not).

File: apps/src/applab/dontMarshalApi.js (hunk 2 above)
Comment (cpirich) [inline apps/src/applab/dontMarshalApi.js:25]: yes, this is just cloned logic from the full version of `apiValidateType` in `commands.js`. But in `dontMarshalApi.js`, I needed to create `dmapiValidateType` that expects the parameters to be unmarshaled interpreter structures.

**Pull request merged**